### ISSUE TITLE
Viewport rect must lie entirely within the current attachment size?

### ISSUE THREAD

**Issue by Richard-Yunchao** [gpuweb/gpuweb#373]
Metal documentation says that "The x/y rectangle of the viewport must lie entirely within the current attachment size".  See https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515527-setviewport?language=objc. 

The current setViewport API in WebGPU doesn't require this. The viewport rect can be bigger than the current attachment size, or it can be set to somewhere outside of the attachment size. 

When I implement this API in Dawn project, I obey the WebGPU's rule and break Metal's requirement, the tests (outside of current attachment size) can pass on Metal backend even the Metal validation layer is on. Looks like Metal's implementation doesn't follow its documentation on this.

Say the attachment's (x, y, w, h) = {0, 0, 4, 4}. The viewport rect is {0, 0, 8, 8}, or {-2, -2, 4, 4} or something like this. 

What do you think, @litherum , @grorg , @JusSn ? Can you verify this with Metal team? BTW, I can find similar statement in Metal documentation for SetScissorRect that the scissor rect must lie entirely inside the attachment. And we do need to obey this rule. 

FYI. @Kangz , @kainino0x , @austinEng . We simply discussed this in Dawn patch.

**Comment by kvark** [comment]
I don't think there is a downside for us to require both viewport and scissor rects to be limited to the render pass attachments size.

**Comment by Richard-Yunchao** [comment]
@kvark , that's an obvious choice. But I am not sure about it. 

I am not an app developer. But when I play with a canvas/image/video or something like that and I resize the window, I can see two different designs are quite common: 1) the rendering result in canvas/image/video will shrink or enlarge as the window changes its size, you can always see the entire rendering result, 2) the rendering result keep the same size, parts of its area are obscured when the window resizes to a smaller one, or the background color (like black) in extra area is showed when the window resizes to a larger one. 

For the 2) situation, I think it might be implemented via changing the back buffer size while keeping the viewport rect when you resize the window. If so, the viewport rect might be outside of the current attachment (or back buffer). Well, if we say viewport rect must lie within the attachment, I guess we can still implement the 2) design (but more difficult?) for webgpu-based canvas if some developers want that design for their apps.

**Comment by kvark** [comment]
@Richard-Yunchao good point!
@jdashg what does WebGL do today to handle resize? Does it care about portable behavior during resize?

**Comment by jdashg** [comment]
WebGL allows oversized viewports, though Firefox does issue a console warning in this case.
Viewport and scissor are not changed automatically on resize.

**Comment by krogovin** [comment]
There are cases where an application may want to use an oversized viewport (I have had one to be honest), but if one the underlying API's says it can't (even it can now), it is likely asking for trouble allowing for it.

**Comment by kainino0x** [comment]
This came up again in https://github.com/gpuweb/cts/pull/276 and https://github.com/gpuweb/gpuweb/pull/850#discussion_r436922968.

We need to figure out what the rules are for both viewports and scissor rectangles.
For viewports I think the two options are:
- Disallow viewports outside the attachment, inconvenient if the attachment changes size.
- Allow, but we can probably only do this if the underlying APIs support it. (IIRC clamping is not semantic-preserving.)

For scissor:
- Disallow scissor outside the attachment; same problem as above?
- Clamp the scissor rectangle to the attachment in the implementation. I believe this is semantic-preserving.

**Comment by Kangz** [comment]
There's quite some [restrictions](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VUID-VkViewport-height-01773) around `setViewport` in Vulkan. Maybe we could restrict it to be inside the attachment.

> Disallow viewports outside the attachment, inconvenient if the attachment changes size.

I'm not sure why this matters because the viewport is set locally to  a render pass, and the attachment size is constant during a render pass.

**Comment by kainino0x** [comment]
I think it would only come up if you're trying to set a fixed-size viewport and you don't handle the case where the window gets resized to smaller than that.

The Vulkan restrictions seem to all be device limits, independent of the attachment size.

**Comment by kainino0x** [comment]
Resolution: Validate setViewport is within the attachment. Continue to investigate the following:
- Flipped viewports (I think we already have an investigation and possibly a decision on this, just need to find it. Maybe #120, in which it's mentioned that this requires VK_KHR_maintenance1.)
- Scissor restrictions (Metal doesn't allow zero-sized scissor rectangle. So if the scissor rect is entirely outside the attachment, we can't just clamp it, because it will end up being zero-sized. We can't no-op draw calls because of vertex shader storage writes.)

**Comment by kvark** [comment]
This might be a little bit silly, but there is an idea to mess with Metal viewport in order to make sure the draw calls are not rasterized. I.e. if the render area has at least 2 pixels, we can make the viewport such that it is fully within one pixel (or no pixels at all?), and the scissor in the another pixel. This doesn't answer the question of what to do if there is only 1x1 rendering :)

**Comment by kainino0x** [comment]
I'm confused now about Metal's setViewport (linked in OP). It appears that Metal has removed the sentence that Yunchao quoted, and it no longer says the viewport has to be within the attachment. I'm assuming this was lifted in a later revision of Metal, but which one?? and is it retroactive?

**Comment by kainino0x** [comment]
It was removed somewhere between [2019-06-18](https://web.archive.org/web/20190618040754/https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515527-setviewport) and now. The documentation page still says "iOS 8.0+, macOS 10.11+" which technically implies it's retroactive, but would need confirmation from @litherum @grorg.

**Comment by Richard-Yunchao** [comment]
IIRC, when I raised that issue, Metal could actually allow that behavior (viewport size can be larger than attachment size), which means that the behavior on Metal conflicts with its documentation at that time. Well, it is understandable, because document may be outdated at that time. And the document was updated at some time.

**Comment by Kangz** [comment]
@kvark if the viewport can be empty then we can make the viewport and scissor disjoint even on 1x1 render targets so that's a good idea to explore.

**Comment by kvark** [comment]
Yeah, I wasn't sure if empty viewport is valid on our target APIs. Needs to be investigated, if it hasn't been already.

**Comment by Kangz** [comment]
Empty viewports are allowed at least in Vulkan 1.1 or Vulkan 1.0 with `VK_KHR_maintenance1`. The spec of that extension says:

> Are viewports with zero height allowed?
>
> RESOLVED: Yes, although they have low utility.

Turns out they are useful!

**Comment by Kangz** [comment]
And the Metal and D3D12 validation allow empty viewports, so it seems empty viewports are possible, which unlocks empty scissors too.

**Comment by litherum** [comment]
Metal has always worked the way the documentation describes now. The previous Metal documentation was always incorrect.

We updated the documentation to match reality.

**Comment by kainino0x** [comment]
Awesome; thanks!

**Comment by kainino0x** [comment]
Assigned to @Kangz who is working on #1138.

**Comment by magcius** [comment]
Should we allow oversized viewports again, given that no backend has an issue with it? I have some (admittedly silly) use cases for it.

**Comment by kainino0x** [comment]
Yes, I think that was the resolution. However we haven't had time to fix it.

**Comment by Kangz** [comment]
We still need to take into account all the restrictions from Vulkan:

> viewportBoundsRange[2] is the [minimum, maximum] range that the corners of a viewport must be contained in. This range must be at least [-2 × size, 2 × size - 1], where size = max(maxViewportDimensions[0], maxViewportDimensions[1]).

> width must be less than or equal to VkPhysicalDeviceLimits::maxViewportDimensions[0]

And more similar restrictions.

**Comment by kainino0x** [comment]
Editors: discussed and the following points came up:
- Unlimited viewport bounds would be more ergonomic than limited viewport bounds (don't have to add special logic when something causes a viewport to go off the edge)
- Limiting viewports to -2..2 like Vulkan would not really improve ergonomics
- Probably can't have unlimited viewport bounds

We think we should delay considering this until we have developer feedback about what they actually would need this for (where they can't easily work around it)

**Comment by kainino0x** [comment]
Oh, there's also discussion in this thread of the scissor rectangle. Right now, we have validation in setScissorRect that requires it to be inside the attachment size (validation error if you go too positive, `[EnforceRange]` exception if you go negative). This is probably not necessary, I think we could replace it remove the validation and change the type to `[Clamp] unsigned long`. It's just nontrivial to be completely certain because of zero-sized rectangle emulation stuff discussed previously.

**Comment by floooh** [comment]
PS:

> Limiting viewports to -2..2 like Vulkan would not really improve ergonomics

I don't agree, this would definitely be good enough. Usually the overall size of a viewport rectangle isn't bigger than the `max(width, height)` of the framebuffer. But such a viewport rectangle may be outside the framebuffer boundary along the 'smaller' dimension (sheesh, this is hard to describe)...

Basically, one way to render 'distortion-free' without messing with projection matrices is to use a viewport rectangle where viewport width and height are identical to the bigger framebuffer dimension, also the viewport remains centered.

Now I can either change the framebuffer width or height as much as I want, but rendering will remain centered and undistorted, becuase the viewport width and height will be identical.

If the viewport size would always be clipped against the framebuffer size (and thus be identical with the framebuffer size), rendering would be squished or stretched when the framebuffer is resized, and that's a problem. Because now I need to use a projection matrix (and matrix multiplication in the vertex shader) to counteract the distortion.

**Comment by kainino0x** [comment]
> Basically, one way to render 'distortion-free' without messing with projection matrices is to use a viewport rectangle where viewport width and height are the same and identical to the bigger framebuffer dimension.

I'd worry this is a bit fragile. On the web it's common to set the canvas size to the size of the window (or some other dynamic value that results from CSS layout). So it's easy to end up with aspect ratios where this would break.

However it's a good point that's a better situation than we're in now.

**Comment by magcius** [comment]
The 2x limit is not for 2x the current viewport size, but for 2x the maximum texture size, which WebGPU mandates is at least 8192, much larger than I would expect most web page sizes to be. I don't think it's likely you'll have viewport sizes below -16384 or above 16384.

**Comment by floooh** [comment]
To quote the relevant parts from the Vulkan documentation (https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceLimits.html)

* viewportBoundsRange[2] is the [minimum, maximum] range that the corners of a viewport must be contained in. This range must be at least [-2 × size, 2 × size - 1], where size = max(maxViewportDimensions[0], maxViewportDimensions[1]). See [Controlling the Viewport](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vertexpostproc-viewport).

...and maxViewportDimensions is defined as:

* maxViewportDimensions[2] are the maximum viewport dimensions in the X (width) and Y (height) dimensions, respectively. The maximum viewport dimensions must be greater than or equal to the largest image which can be created and used as a framebuffer attachment. See [Controlling the Viewport](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vertexpostproc-viewport).

...and the following note regarding viewportsBoundsRange:

> The intent of the viewportBoundsRange limit is to allow a maximum sized viewport to be arbitrarily shifted relative to the output target as long as at least some portion intersects. This would give a bounds limit of [-size + 1, 2 × size - 1] which would allow all possible non-empty-set intersections of the output target and the viewport. Since these numbers are typically powers of two, picking the signed number range using the smallest possible number of bits ends up with the specified range.

We should probably check if this definition is the same all the way back to Vulkan 1.0.

**Comment by kainino0x** [comment]
Ah, sorry, I misremembered or never correctly understood what the -2x..2x meant. Thanks for the clarification.

**Comment by kainino0x** [comment]
Another request for this feature in https://github.com/gpuweb/gpuweb/discussions/4335

**Comment by kdashg** [comment]
Can we just polyfill this?

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-12-04#maybe-viewport-rect-must-lie-entirely-within-the-current-attachment-size-373">GPU Web WG 2024-12-04 Atlantic-time</a></summary>

* KG: Polyfill?
* KN: Shouldn't even need polyfilling, all backends support it
* BJ: Except Vulkan has detailed limits for this.
* KN: IIUC Vulkan requires those limits to be at least 2*max texture size, and that would work on other backends, so can just make that the hard limit. (not introduce a new limit).
* **KG: Ok! I think we’re just waiting on a PR to relax this [somewhat].**
* BJ: Will likely pick it up
* AB: What about zero-area viewport? Requires Vulkan 1.1 or maintenance1.
* KN: We already require 1.1/maintenance1.
* BJ: Use case?
* AB: Unreal Engine uses it in multiview stuff for turning off a view, or something.

</details>

**Comment by floooh** [comment]
I'm very happy that this got resolved :)
